Add support querying IP routes

cathay4t opened the pull request:
The example output would be:

```yaml
  - destination: 0.0.0.0/0
    metric: 100
    next-hop-address: 192.0.2.251
    next-hop-iface: ens10
    origin: dhcp
    status: up
    table-id: 254
    table-name: main
  ipv6:
  - destination: ::/0
    metric: 100
    next-hop-address: fe80::be2b:a3e2:2b5a:2e7d
    next-hop-iface: ens10
    origin: router-advertisement
    status: up
    table-id: 254
    table-name: main
```

Updated nmstatectl to allow filter routes base on 'next-hop-iface'.


**Depending on PR #246**

tyll commented on libnmstate/netinfo.py:
```diff
@@ -36,6 +36,9 @@ def show(include_status_data=False):
     if include_status_data:
         report['capabilities'] = capabilities()
 
+    report[Constants.ROUTES] = nm.ipv4.get_route_info()
```
`report[Constants.ROUTES] = nm.ipv4.get_route_info() + nm.ipv6.get_route_info()` could work as well and be easier to read.

cathay4t replied:
Done.

tyll commented on libnmstate/nm/ipv4.py:
```diff
@@ -15,9 +15,16 @@
+
+IPV4_DEFAULT_GATEWAY_NETWORK = '0.0.0.0/0'
+MAIN_ROUTE_TABLE = 'main'
+MAIN_ROUTE_TABLE_INT = 0
```
This should be `ID` instead of `INT` I guess.

cathay4t replied:
Done. 
Changed to:

```
MAIN_ROUTE_TABLE_NAME = 'main'
MAIN_ROUTE_TABLE_ID = 0
```

tyll commented on libnmstate/nm/ipv4.py:
```diff
@@ -119,3 +126,95 @@ def get_ip_profile(active_connection):
+                next_hop = ''
+            metric = int(route.get_metric())
+            route_entry = {
+                'status': 'up',
```
We started using constants for field names, therefore it would be great to use them here as well. See the OVS code changes for example.

cathay4t replied:
OK. Will do.

cathay4t replied:
Done.

tyll commented on libnmstate/nm/ipv4.py:
```diff
@@ -119,3 +126,95 @@ def get_ip_profile(active_connection):
+
+        devs = active_connection.get_devices()
+        iface_name = None
+        # If we have master device, we use it, else we use first device.
```
This might be valid for Nmstate but in general I believe a profile could be active for multiple interfaces now. Not sure what kind of static routes this would result in.

cathay4t replied:
Let's address that in another PR.

tyll commented on libnmstate/nm/ipv4.py:
```diff
@@ -119,3 +126,95 @@ def get_ip_profile(active_connection):
+                'table-id': table_id,
+                'table-name': iplib.get_route_table_name(table_id),
+                'destination': dst,
+                'next-hop-iface': iface_name,
```
It will be interesting how we will figure out if the interface was explicitly requested by the user or not here.. @EdDev will be good if you can share your thoughts in the future.

EdDev replied:
We just report reality, which mean it is always included in the report. If the user specified it or not, we do not really care.

Do you see a problem with this?

tyll replied:
If you want to allow users to not specify the interface but it appears in the output, they will always fixate the interface in the route, i.e. `apply(show())` will not be a no-op.

tyll commented on libnmstate/nm/ipv6.py:
```diff
@@ -16,12 +16,19 @@
 
+IPV6_DEFAULT_GATEWAY_NETWORK = '::/0'
+MAIN_ROUTE_TABLE = 'main'
+MAIN_ROUTE_TABLE_INT = 0
```
s/INT/ID please

tyll commented on libnmstate/nm/ipv6.py:
```diff
@@ -158,3 +165,110 @@ def get_ip_profile(active_connection):
+        if not ip_cfg:
+            continue
+
+        devs = active_connection.get_devices()
```
This code seems to be the same as in the IPv4 module. Would be great to move it to a shared function.

cathay4t replied:
Edy seems don't allow me to create a `route.py` to hold this function as he want to align with NM data structure in `libnmstate/nm` folder.

EdDev replied:
You can create such a module, but I think you should expose it through the ipX modules as well, just to keep a sync with nm structure.

Maybe at the ipX modules detect what nm object is needed, then pass it with the data to a route module. So, common stuff in route and specific (6/4) in ipX.

tyll replied:
If the module can be generic for IP a name like `ip.py` makes more sense. Cannot remember that people use `ipX` to say that it can be any version of IP. It sounds more like IPX.

cathay4t replied:
I have moved the codes to `route.py`, the `ipv4.py` and `ipv6.py` only provide the function reference to `route.py` to use the unified codes.

tyll commented on libnmstate/nm/ipv6.py:
```diff
@@ -158,3 +165,110 @@ def get_ip_profile(active_connection):
+    return ret
+
+
+def _should_skip(dst, skips):
```
This can be shared withthe IPv4 module.

cathay4t replied:
Done.

tyll commented on libnmstate/nm/ipv6.py:
```diff
@@ -158,3 +165,110 @@ def get_ip_profile(active_connection):
+
+def _should_skip(dst, skips):
+    '''
+    Check whether specified route destination should be skipped for unicast
```
The docstring is not really matching what the method does because it does not have the information about direct routes or link-local networks by itself.

tyll commented on libnmstate/nm/ipv6.py:
```diff
@@ -158,3 +165,110 @@ def get_ip_profile(active_connection):
+    return False
+
+
+def _get_route_table_id(active_connection, ip_cfg):
```
Most of this code could be shared with the IPv4 module... The `ip_cfg` argument does not seem to be used currently because it reassigned without any conditions.

cathay4t replied:
New patch uploaded, a hack(passing the function reference) has been used to let `route.py` don't check on IP family.

cathay4t commented:
New patch uploaded.

Changes:
 * Using merged route schema.
 * New script file `route.py` to hold the shared codes between ipv4 and ipv6.

tyll commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,110 @@
+    return 'static'
+
+
+def _get_route_table_id(active_connection, ip_cfg, get_ip_profile_func):
```
IMHO you could just pass the ip_profile instead of the function as an argument. Also the ip_cfg argument does not seemt to be used.

cathay4t replied:
Done. Thanks.

tyll commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,110 @@
+                      is_auto_route_func, auto_origin_value):
+    if dst not in static_routes and is_auto_route_func(active_connection):
+        return auto_origin_value
+    return 'static'
```
This should be a constant

cathay4t replied:
Done.

cathay4t commented:
TODO:
 * Change file name to constants.
 * Investigate how static routes on profile with multiple interfaces.

EdDev commented on libnmstate/iplib.py:
```diff
@@ -14,12 +14,85 @@
+    """
+    a = ipaddress.ip_network(six.u(network_a), strict=False)
+    b = ipaddress.ip_network(six.u(network_b), strict=False)
+    if hasattr(a, 'subnet_of'):
```
Please use a module `compat` to abstract this.
See `compat.py` in the tests.

cathay4t replied:
Done.

EdDev commented on libnmstate/iplib.py:
```diff
@@ -14,12 +14,85 @@
             prefix >= _IPV6_LINK_LOCAL_NETWORK_PREFIX_LENGTH)
+
+
+def _is_subnet_of(a, b):
```
Please keep the caller below the callee so the details will always be below.
This on need to be placed below `is_subnet_of`.

cathay4t replied:
Done.

cathay4t commented:
> Please split the changes done in the `nm` subpackage from the rest (I mean a separate PR). The changes should come with integration tests coverage in `tests/integration/nm/`.

I don't like the idea of using `tests/integration/nm/` as it does not cover real user case. I will expand the `dhcp_test` for these route query codes.

cathay4t commented:
New patch uploaded.

Changes:
 * Use constants `LinuxRoute, LinuxRouteOrigin, Constants.ROUTES` for all the key names for linux routes.
  * Update the integration test `dhcp_test.py` to use `netinfo.show()` for route information query.

EdDev commented:
> I don't like the idea of using tests/integration/nm/ as it does not cover real user case. I will expand the dhcp_test for these route query codes.

The NM tests are a sane replacement for unit tests in this subpackage. Unit testing it requires too many mocks, making it unmaintainable.
These nm subpackage tests are focused and show how to use the nm subpackage.

cathay4t commented:
New patch uploaded.

Changes:
 * Use `libnmstate/compat.py` file to abstract the `subnet_of()` function for python2, 3,6-.

EdDev commented on libnmstate/compat.py:
```diff
@@ -0,0 +1,34 @@
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+
+def subnet_of(a, b):
```
I think the way it should go is:
```
import ipaddress
import iplib

if six.PY2:
    ipaddress.IPv4Network.subnet_of = iplib.subnet_of
```

But I am missing something here.
`subnet_of` exists in the python-ipaddress package (or directly from pip) which is a backport of py3 ipaddress.
If we depend on that one, we will not need any hack.
Or do I miss something.

cathay4t replied:
The `subnet_of()` is new in version python 3.7, above line is for python 3.6-.

cathay4t replied:
> the user should import ipaddress and use it the same way it is documented in the docs.

Who is the `user` here? It's not a API, for internal use, internal function call should be through `iplib.is_subnet_of` which hide the detail of using `ipaddress`.

EdDev replied:
The user is anyone, internal or external. There is no difference between internal and external API except for the flexibility to change the internal one more frequently.

A developer should not think about this stuff at all. If this is already introduced in 3.7 under ipaddress, we can make it look as if all versions have this available and when we reach 3.7 only support, we can drop the patching/wrappers.

cathay4t commented:
> Please note that the added code has not been covered by any unit or integration tests.

The test is covered by `dhcp_test.py`.

EdDev commented on libnmstate/nm/ipv4.py:
```diff
@@ -119,3 +122,19 @@ def get_ip_profile(active_connection):
+def get_route_info():
+    client = nmclient.client()
+    skip_networks = [iplib.IPV4_LINK_LOCAL_NETWORK]
+    func = {
```
This looks like an attempt to push everything into one handler instead of splitting things up into multiple handlers which are called from here to process the needed.
Sometimes, especially when there is a complex logic, this makes sense, but here I'm not sure it is needed in this case. The first smell will be that this func does too many things.
Even if it is needed, using a dict to collect the funcs is not that nice. It will be nicer to use a direct argument variable.

cathay4t replied:
Thanks for the suggest, I will try find a better way.

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,147 @@
+        iface_name = _get_iface_name(active_connection)
+        if not iface_name:
+            logging.warning(
+                'Got connection {} has not interface name'.format(
```
Active connection {} has no interface name

Is this an error or a supported option? In what cases this may happen?

cathay4t replied:
No idea, I never got this. The code is ensure we could get a warning when strange things happens.

EdDev replied:
I guess the question is: Why not explode?
This is especially true if we do not know, it makes it an unexpected case and we need to detect it somehow.
At the minimum I would log the exception itself (logging.exception), then consider if to ignore the error or make a noise. I tend here to explode and let the caller/user/client try again and open a bug.

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,147 @@
+
+def _get_skip_networks(ip_cfg):
+    """
+    Currently libnm cannot manage the direct route of interface's LAN:
```
... cannot manage directed connected routes.
Example:
   eth1 with an IP address ...
   has the following direct route: ...

But I do not understand why you need to hide it.

cathay4t replied:
As libnm cannot edit on direct route, showing them seems like a noise to me.

EdDev commented:
> The subnet_of addition should be transparent to the code that uses it as if it uses ipaddress from py3.7. You could add an ipaddress module as an extension to the original one and do the needed to add the missing functionality. This fits under its own commit.

Please see #303

cathay4t commented:
I will create a demo for `show_running_state()` and `show_saved_config()`, then we could revisit this PR.

cathay4t commented:
@EdDev  @tyll 
New patch uploaded.
Changes: 
 * Schema changed: split routes into running routes and in-config routes.
 * Rebased to master to use the ipaddress in `libnmstate/compat.py`.

EdDev commented on libnmstate/schemas/operational-state.yaml:
```diff
@@ -435,3 +419,23 @@ definitions:
+        enum:
+          - absent
+      table-name:
+        type: string        # empty for unknown
```
I would prefer `unknown` to be expressed by not specifying it in the config.
It is also not common to have comments in schemas and an empty string is self explanatory.

tyll replied:
IMHO the empty string is not self explanatory because it could mean both that the table has no name or that the name of the table is unknown. But why can the table name be unknown?

cathay4t replied:
`table-name` has been removed.

EdDev commented on libnmstate/schemas/operational-state.yaml:
```diff
@@ -435,3 +419,23 @@ definitions:
+      table-name:
+        type: string        # empty for unknown
+      table-id:
+        type: integer       # -1 means unknown
```
It cannot be unknown as any route belongs to some table.

cathay4t replied:
Fixed by getting route table id from `NM.IPRoute.get_attribute()`.

EdDev commented on tests/lib/schema_validation_test.py:
```diff
@@ -64,26 +64,28 @@
+        'running': [
+            {
+                'table-name': '',
+                'table-id': -1,
```
It cannot be negative as any route belongs to some table.

cathay4t replied:
Fixed.

EdDev replied:
Still negative.

EdDev commented on libnmstate/schema.py:
```diff
@@ -41,9 +41,22 @@ class Interface(object):
     MTU = 'mtu'
 
 
+class LinuxRoute(object):
```
Any reason you prefixed it with `Linux`?

cathay4t replied:
Changed to `Route`.

EdDev commented on libnmstate/nm/ipv4.py:
```diff
@@ -119,3 +121,17 @@ def get_ip_profile(active_connection):
+
+def get_route_running():
+    client = nmclient.client()
+    skip_networks = [iplib.IPV4_LINK_LOCAL_NETWORK]
```
As we now split the config and running, there is no need to hide any routes.

cathay4t replied:
Done.

EdDev commented on libnmstate/nm/ipv6.py:
```diff
@@ -158,3 +163,24 @@ def get_ip_profile(active_connection):
+
+def get_route_running():
+    client = nmclient.client()
+    skip_networks = [iplib.IPV6_LINK_LOCAL_NETWORK,
```
As we now split the config and running, there is no need to hide any routes.

cathay4t replied:
Done.

EdDev commented on libnmstate/nm/ipv6.py:
```diff
@@ -158,3 +163,24 @@ def get_ip_profile(active_connection):
+        skip_networks)
+    for route in routes:
+        if route[LinuxRoute.METRIC] == 0:
+            # Kernel will convert 0 to IPV6_DEFAULT_ROUTE_METRIC.
```
Isn't this a NM bug then?

cathay4t replied:
That's from NM document.

cathay4t replied:
Note that for IPv6, the kernel accepts zero (0) but coerces it to 1024 (user default). Hence, setting this property to zero effectively mean setting it to 1024. For IPv4, zero is a regular value for the metric.

EdDev replied:
Sorry, I do not understand why this is not an NM bug.
I would expect NM to return 1024 and not 0, because 0 is not a valid number when reading it from the kernel.
If someone sets 0 in the NM config, it should wither fail or save it as 1024 implicitly (like the kernel does).

cathay4t replied:
I fixed our code to only check config routes for zero IPv6 metric and tested.

Bug reported at: https://bugzilla.redhat.com/show_bug.cgi?id=1692911

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,150 @@
+
+def get_running(client, get_ip_cfg_func, skip_networks):
+    ret = []
+    for active_connection in client.get_active_connections():
```
You could have passed the ip_cfg list instead of passing the client and the function pointer.
Not a must, but as only the ipvX modules should know about this function anyway, they could just do the calling there and pass the result down to here:
```
acs_and_ip_confs = [(ac, nmclient.NM.ActiveConnection.get_ip4_config(active_connection)) for ac in client.get_active_connections()]
get_running(acs_and_ip_confs)
```

If you want the actual execution to happen at the iteration point and save some memory, you can build it as a generator.

cathay4t replied:
Done.

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,150 @@
+                    active_connection.get_id()))
+            continue
+        skip_networks.extend(_get_skip_networks(ip_cfg))
+        table_id = -1
```
I guess you can use the main table here (254).
255     local
254     main
253     default

cathay4t replied:
Done.

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,150 @@
+            continue
+        skip_networks.extend(_get_skip_networks(ip_cfg))
+        table_id = -1
+        # TODO: Waiting libnm to provide route table id for running state
```
A workarround could be to check it using iproute2 or check if it is in the config and copy it from there.
But it can be "fixed" in a separate PR.

As it stands now, it will not be useful for clients like VDSM which use the tables extensively.

cathay4t replied:
Fixed.

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,150 @@
+        skip_networks.extend(_get_skip_networks(ip_cfg))
+        table_id = -1
+        # TODO: Waiting libnm to provide route table id for running state
+        for libnm_route in ip_cfg.get_routes():
```
nit: I would use `nm` as a prefix and not `libnm`. The whole nm subpackage is about libnm.

cathay4t replied:
Changed to `nm_route`.

cathay4t commented on libnmstate/iplib.py:
```diff
@@ -14,12 +14,66 @@
+                match = regex.match(line.strip())
+                if match:
+                    return match.groups()[0]
+    except Exception as e:
```
Above code has been removed as we removed the `table-name` property.

cathay4t commented on libnmstate/nm/ipv6.py:
```diff
@@ -158,3 +163,24 @@ def get_ip_profile(active_connection):
+
+def get_route_running():
+    client = nmclient.client()
+    skip_networks = [iplib.IPV6_LINK_LOCAL_NETWORK,
```
We don't skip anymore.

cathay4t commented on libnmstate/nm/ipv4.py:
```diff
@@ -119,3 +121,17 @@ def get_ip_profile(active_connection):
+
+def get_route_running():
+    client = nmclient.client()
+    skip_networks = [iplib.IPV4_LINK_LOCAL_NETWORK]
```
We don't skip anymore.

cathay4t commented on libnmstate/nm/ipv6.py:
```diff
@@ -158,3 +163,24 @@ def get_ip_profile(active_connection):
+        nmclient.NM.ActiveConnection.get_ip6_config,
+        skip_networks)
+    for route in routes:
+        if route[LinuxRoute.METRIC] == 0:
```
I failed to find constants in NM for it. Only document said:
```
Note that for IPv6, the kernel accepts zero (0) but coerces it to 1024 (user default). Hence, setting this property to zero effectively mean setting it to 1024. For IPv4, zero is a regular value for the metric.
```

cathay4t commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,150 @@
+
+def get_running(client, get_ip_cfg_func, skip_networks):
+    ret = []
+    for active_connection in client.get_active_connections():
```
Done.

cathay4t commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,150 @@
+
+
+def get_running(client, get_ip_cfg_func, skip_networks):
+    ret = []
```
Done.

cathay4t commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,150 @@
+        skip_networks.extend(_get_skip_networks(ip_cfg))
+        table_id = -1
+        # TODO: Waiting libnm to provide route table id for running state
+        for libnm_route in ip_cfg.get_routes():
```
The `ip_cfg.props.routes` is the same with `ip_cfg.get_routes()`.

I will try to merge them into single function if `route_table` is also the same.

cathay4t replied:
The route table ID is handle differently. Still working to merge them.

cathay4t replied:
Merged to `nm.route.get_routes()`.

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,111 @@
+
+def get_running(acs_and_ip_confs):
+    ret = []
+    for (active_connection, ip_cfg) in acs_and_ip_confs:
```
nit: I do not think you need parenthesis

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,111 @@
+
+def get_config(acs_and_ip_profiles):
+    ret = []
+    for (active_connection, ip_profile) in acs_and_ip_profiles:
```
nit: The parenthesis are redundant.

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,111 @@
+    for (active_connection, ip_profile) in acs_and_ip_profiles:
+        iface_name = _get_iface_name(active_connection)
+        if not iface_name:
+            logging.warning(
```
You should probably raise like in `get_running`.

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,112 @@
+    """
+    Return interface name for active_connection, return None if error.
+    """
+    # If we have master device, we use it, else we use first device.
```
Why is this correct? Could you explain the scenario?
If this is an ac of a bond slave, that slave cannot have any IP on it and therefore routes through it are irrelevant.
Same for a port of a bridge or even a base of a vlan (in the vlan case it is a totally different iface).

cathay4t replied:
I will take a look tomorrow.

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,112 @@
+    dst = '{ip}/{prefix}'.format(
+        ip=nm_route.get_dest(), prefix=nm_route.get_prefix())
+    next_hop = nm_route.get_next_hop()
+    if not next_hop:
```
nit: `next_hop = nm_route.get_next_hop() or ''`

cathay4t commented:
New patch uploaded.

Changes:
 * Removed duplicate code in `nm.route` to just have single function for both query running and config routes.
 * Moved the IPv6 zero metric fixing code from running routes to config routes as that only happens in config.
 * Renamed the constant `LinuxRoute` to `Route`.
 * Removed the `table-name` property from schema.

EdDev commented:
> Removed duplicate code in nm.route to just have single function for both query running and config routes.

I consider this change bad because it removed duplication by adding type checking all over.
I prefer duplication over having type based conditions inside that function.

Sorry, but the profile setting and active_connection  properties are not sharing any common thing, the fact that you retrieve similarly info from them is not duplication.

I hope you can revert back to the previous version.

cathay4t commented:
Reverted to duplicate code version: `nm.route.get_running()` and `nm.route.get_config()`.

cathay4t commented:
New patch uploaded.

Changes:
 * Don't show route of slaves when route is on mater interface.
 * Check the existence of routes before doing interface name query.

EdDev commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,124 @@
+    Return interface name for active_connection, return None if error.
+    """
+    devs = active_connection.get_devices()
+    if devs:
```
nit: `return devs[0].get_iface() if dev else None`

EdDev commented on requirements.txt:
```diff
@@ -11,3 +11,4 @@ PyGObject
 PyYAML
 setuptools
 six
+ipaddress;python_version<"3.3"
```
Already exists in line 8

EdDev commented on tests/ctl/nmstatectl_test.py:
```diff
@@ -24,33 +24,54 @@
+EMPTY_ROUTE_JSON_STATE = """"{ROUTES}": {{
+        "{ROUTE_CONFIG}": [],
+        "{ROUTE_RUNNING}": []
+    }}""".format(ROUTES=Constants.ROUTES,
```
If it is a yaml string, I prefer to have it straight clean so it can be copy/pasted.
The point if a yaml in the test is readability.

If we need to build it from constants, then I think it will be better to use a straight dict and not yaml.

cathay4t commented:
New patch uploaded.

Changes:
 * Fix all the lints suggested by @EdDev 
 * Add support of multiple route table id as https://bugzilla.redhat.com/show_bug.cgi?id=1436531 suggested and also tested manually. (Will add test in route editing PR)

tyll commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,129 @@
+        # table id for static routes and auto routes. But each static route can
+        # still specify route table id.
+        for nm_route in ip_profile.props.routes:
+            table = nm_route.get_attribute('table')
```
You could change ` _get_per_route_table_id(nm_route)` to also take an optional default ID argument and then re-use it here, too.

cathay4t replied:
Sure. Will do.

tyll commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,129 @@
+            continue
+        iface_name = _get_iface_name(active_connection)
+        if not iface_name:
+            raise NmstateInternalError(
```
The exception could be raised in `_get_iface_name()` directly. Also it seems to me that it would make more sense to pass the interface name instead of the active connection to get_running() and get_config().

cathay4t replied:
If that activeconnection has no routes on it, there is no need to query its interface name or even raise exception for unknown iface name.
But clearly I do more investigation on this. How about we address this in another PR.

tyll commented on libnmstate/nm/ipv6.py:
```diff
@@ -158,3 +163,33 @@ def get_ip_profile(active_connection):
     return None
+
+
+def get_route_running():
```
IMHO it would be more natural to use `get_running_routes()` instead of reversing the words. Also for `get_configured_routes()`.

cathay4t commented:
New patch uploaded.

Changes:
 * Remove the leftover of `table-name` property.
 * Updated the `_get_per_route_table_id()` to support passing default table id to remove duplication.

tyll commented on libnmstate/nm/route.py:
```diff
@@ -0,0 +1,125 @@
+
+def _get_per_route_table_id(nm_route, default_table_id):
+    table = nm_route.get_attribute('table')
+    return table.get_uint32() if table else default_table_id
```
This seems to result in `table-id: !!python/long '255'` in the yaml, so adding a `int()` here makes sense.

cathay4t replied:
Added the `int()`.

tyll commented:
The nmstatectl edit command needs to allow editing only routes:

```
./ncl edit bond99
ERROR: Invalid state: should contain "interfaces" entry.
Try again? [y,n]:
y - yes, start editor again
n - no, throw away my changes
> 

```

EdDev commented:
ci test please

cathay4t commented:
New patch uploaded.

Changes:
 * Use `int()` to convert table id u32 to python integer.
 * Allow editing routes only in nmstatectl.

cathay4t commented:
ci test please

EdDev commented:
ci test please

EdDev commented:
ci test please

EdDev commented:
ci test please
